client/post-editor/media-modal/index.jsx

```diff
@@ -174,6 +180,11 @@ export const EditorMediaModal = React.createClass( {
 
 		MediaActions.delete( site.ID, toDelete );
 		analytics.mc.bumpStat( 'editor_media_actions', 'delete_media' );
+		const mediaIds = map( toDelete, 'ID' );
+		this.props.deleteMedia( site.ID, mediaIds );
+		if ( includes( mediaIds, siteIconId ) ) {
+			this.props.resetSiteIcon( site.ID );
+		}
```

Comment:
Yeah, I had a hard time coming to a solution here. It's complicated by the fact that the data exists in many places (media objects themselves, a media ID as site setting, and a media's URL as the site icon property). But you're right that any part of the app which deletes a media would then need to be concerned with resetting the site icon, which is not great.

Similar to what I suggested at https://github.com/Automattic/wp-calypso/pull/10032#issuecomment-267093132, perhaps a middleware (the API middleware?) is the right place for these kinds of side effects. While we're not quite at the point where we want to substitute the API call for deleting media, from there we could at least check whether the media to be deleted matches the current site option.

Or, maybe since it's not related to network calls on its own, it could exist as a standalone middleware, or combined with other media middleware suggested in #9706.

cc @dmsnell in case you have thoughts on the role of middlewares / API middleware